micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala

```diff
@@ -0,0 +1,113 @@
+package zio.metrics.connectors.micrometer
+
+import java.lang.{Double => JDouble}
+import java.time.Instant
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicLong
+
+import scala.jdk.CollectionConverters.IterableHasAsJava
+
+import zio.{Unsafe, ZIO}
+import zio.metrics.{MetricKey, MetricKeyType, MetricListener}
+import zio.metrics.MetricKeyType.{Counter, Frequency, Gauge}
+
+import io.micrometer.core.instrument.{DistributionSummary, MeterRegistry, Tag}
+
+private[micrometer] class MicrometerMetricListener(meterRegistry: MeterRegistry) extends MetricListener {
+  private lazy val gauges: ConcurrentHashMap[MetricKey[Gauge], AtomicDouble] =
+    new ConcurrentHashMap[MetricKey[Gauge], AtomicDouble]()
+
+  override def updateHistogram(
+    key: MetricKey[MetricKeyType.Histogram],
+    value: Double,
+  )(implicit unsafe: Unsafe,
+  ): Unit =
+    DistributionSummary
+      .builder(key.name)
+      .tags(micrometerTags(key.tags).asJava)
+      .serviceLevelObjectives(key.keyType.boundaries.values.filter(_ > 0): _*) // micrometer prohibits <= 0 slo values
+      .register(meterRegistry)
+      .record(value)
+
+  override def updateGauge(key: MetricKey[Gauge], value: Double)(implicit unsafe: Unsafe): Unit =
+    gauges
+      .computeIfAbsent(
+        key,
+        _ =>
+          meterRegistry.gauge(
+            key.name,
+            micrometerTags(key.tags).asJava,
+            AtomicDouble.make(0),
+            (v: AtomicDouble) => v.get(),
+          ),
+      )
+      .set(value)
+
+  override def updateFrequency(key: MetricKey[Frequency], value: String)(implicit unsafe: Unsafe): Unit =
+    meterRegistry
+      .counter(
+        key.name,
+        (micrometerTags(key.tags) ++ Iterable(Tag.of("bucket", value))).asJava,
+      )
+      .increment()
+
+  override def updateSummary(
+    key: MetricKey[MetricKeyType.Summary],
+    value: Double,
+    instant: Instant,
+  )(implicit unsafe: Unsafe,
+  ): Unit =
+    DistributionSummary
+      .builder(key.name)
+      .tags(
+        (micrometerTags(key.tags) ++ Iterable(Tag.of("error", key.keyType.error.toString))).asJava,
+      )
+      .distributionStatisticBufferLength(key.keyType.maxSize)
+      .distributionStatisticExpiry(key.keyType.maxAge)
+      .publishPercentiles(key.keyType.quantiles: _*)
+      .percentilePrecision(3) // should this be added to config?
+      .register(meterRegistry)
+      .record(value)
+
+  override def updateCounter(key: MetricKey[Counter], value: Double)(implicit unsafe: Unsafe): Unit =
+    meterRegistry
+      .counter(key.name, micrometerTags(key.tags).asJava)
+      .increment(value)
```

Comment:
I will try to explain the above using the example of a counter

```suggestion
    {
    // add config: zio.metrics.connectors.MetricsConfig field to the constructor

    val name = key.name.replaceAll("-", ".").trim
    val description = key.tags.find(_.key == config.descriptionKey).map(_.value)
    val tagsWithoutDescription = key.tags.filterNot(_.key == config.descriptionKey).map(micrometerTags)

    Counter
      .builder(name)
      .description(description.orNull)
      .tags(tagsWithoutDescription.asJava)
      .register(registry)
      .increment(value)
  }
```